R/derive_param_tte.R

```diff
@@ -0,0 +1,283 @@
+    data[[i]] <- sources[[i]]$dataset %>%
+      filter_if(sources[[i]]$filter) %>%
+      filter_extreme(
+        order = vars(!!date),
+        by_vars = subject_keys,
+        mode = mode,
+        check_type = "none"
+      )
+    # add date variable and accompanying variables
+    if (is.Date(data[[i]][[as_string(date)]])) {
+      data[[i]] <- transmute(data[[i]],
+                             !!!subject_keys,
+                             !!!sources[[i]]$set_values_to,
+                             CNSR = sources[[i]]$censor,
+                             ADT = !!date)
+    } else if (is.instant(data[[i]][[as_string(date)]])) {
+      data[[i]] <- transmute(data[[i]],
+                             !!!subject_keys,
+                             !!!sources[[i]]$set_values_to,
+                             CNSR = sources[[i]]$censor,
+                             ADT = date(!!date))
+    } else {
+      data[[i]] <- transmute(
+        data[[i]],
+        !!!subject_keys,
+        !!!sources[[i]]$set_values_to,
+        CNSR = sources[[i]]$censor,
+        ADT = convert_dtc_to_dt(!!date,
+                                date_imputation = "first")
+      )
+    }
+  }
+
+  # put all source data into one dataset and select first or last date per subject
+  data %>%
+    bind_rows() %>%
+    filter(!is.na(ADT)) %>%
+    filter_extreme(
+      by_vars = subject_keys,
+      order = vars(ADT),
+      mode = mode,
+      check_type = "none"
+    )
+}
+
+#' Create an `tte_source` object
+#'
+#' The `tte_source` object is used to define events and possible censorings as
+#' input for the `derive_param_tte()` function.
+#'
+#' @param dataset The source dataset
+#'
+#' @param filter An unquoted condition for selecting the observations from
+#'   `dataset` which are events or possible censoring time points.
+#'
+#' @param date A variable providing the date of the event or censoring. A date,
+#'   a datetime, or a character variable containing ISO 8601 dates can be
+#'   specified. An unquoted symbol is expected.
+#'
+#' @param censor Censoring value
+#'   CDISC strongly recommends using `0` for events and positive integers for
+#'   censoring.
+#'
+#' @param set_values_to A named list returned by `vars()` defining the
+#'   variables to be set for the event or censoring, e.g. `vars(EVENTDESC =
+#'   “DEATH”, SRCDOM = “ADSL”, SRCVAR = “DTHDT”)`. The values must be a symbol,
+#'   a character string, a numerig value, or `NA`.
+#'
+#' @author Stefan Bundfuss
+#'
+#' @keywords source_specifications
+#'
+#' @export
+#'
+#' @return An object of class "tte_source".
+tte_source <- function(dataset,
+                       filter = NULL,
+                       date,
+                       censor = 0,
+                       set_values_to = NULL) {
+  out <- list(
+    dataset = assert_data_frame(dataset),
+    filter = assert_filter_cond(enquo(filter), optional = TRUE),
+    date = assert_symbol(enquo(date)),
+    censor = assert_integer_scalar(censor),
```

Comment:
I am not sure if and how we should check the values (see my message in the admiral_testing_01 channel).

I set the default value to `0` because I expect you need to define much more events than censorings.